lib/spack/spack/solver/error_messages.lp

```diff
@@ -0,0 +1,216 @@
+% Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+% Spack Project Developers. See the top-level COPYRIGHT file for details.
+%
+% SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+%=============================================================================
+% This logic program adds detailed error messages to Spack's concretizer
+%=============================================================================
+
+#program error_messages.
+
+% Create a causal tree between trigger conditions by locating the effect conditions
+% that are triggers for another condition. Condition2 is caused by Condition1
+condition_cause(Condition2, Condition1, ID) :-
```

Comment:
I think we should rewrite as:
```suggestion
condition_cause(EffectID, CauseID, PackageID) :-
```
I'm also wondering if we should pass the full `node(PackageID, Package)` for clarity (instead of only `PackageID`).

Reply:
Yes, that's probably something to address in the future. Off the top of my head I'm not thinking of another attribute name for which that would occur, but I expect we will find more.

Reply:
It's probably these attributes:
```
node_platform = fn.attr("node_platform_set")
node_os = fn.attr("node_os_set")
node_target = fn.attr("node_target_set")
variant_value = fn.attr("variant_set")
node_compiler = fn.attr("node_compiler_set")
node_compiler_version = fn.attr("node_compiler_version_set")
node_flag = fn.attr("node_flag_set")
```